src/DynamoCore/Scheduler/UpdateGraphAsyncTask.cs

```diff
@@ -59,6 +94,31 @@ internal bool Initialize(EngineController controller, WorkspaceModel workspace)
                 TargetedWorkspace = workspace;
 
                 ModifiedNodes = ComputeModifiedNodes(workspace);
+
+                // Detect cyclic dependencies in graph
+                var cyclicNodes = new List<NodeModel>();
+
+                foreach (var node in ModifiedNodes)
```

Comment:
hmm, so this appears that it will start multiple DFS searches from each modified node - on opening a new graph all nodes will be modified so this has the possibility of recomputing the same information multiple times -and could have a performance like `O(n(n+v))` if I've got that right.

can't each invocation of DFS share the visited data structure? If one DFS has already computed this subtree of the graph whats the reason to recompute it?

Reply:
@mjkkirschner, you're right. I made a mistake earlier. I was initializing the `visited` and `recursion` lists for each modified node. I have now moved their initialization outside the `for` loop so it's done only once. That way if a node is already visited it will be skipped. The overall complexity of this algo is O(v+e). The reason why I'm running this on all nodes as there could be disconnected portions of the graph and I am trying to collect all nodes involved in a cycle and report warnings on those. 

The good thing is this will run on all nodes only the first time a graph is opened and executed but subsequently, it will only run on the modified nodes for each delta execution.